File: BaekjoonOnlineJudge/DFS/DFS와BFS (1260)/Solution.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

bool visitedDFS[1001];
bool visitedBFS[1001];
vector<int> graph[1001];

void dfs(int x) {
	visitedDFS[x] = true;
	cout << x << ' ';
	for (int i = 0;i < graph[x].size();i++) {
		int y = graph[x][i];
		if (!visitedDFS[y]) dfs(y);
	}
}

void bfs(int start) {
	queue<int> q;
	q.push(start);

	visitedBFS[start] = true;

	while (!q.empty()) {
		int x = q.front();
		q.pop();
		cout << x << ' ';
		
		for (int i = 0;i < graph[x].size();i++) {
			int y = graph[x][i];
			if (!visitedBFS[y]) {
				q.push(y);
				visitedBFS[y] = true;
			}
		}
	}
}

int main() {
	int N,M,V;

	cin >> N >> M >> V;

	for (int i = 1;i <= M;i++) {
		int a, b;
		cin >> a >> b;
		graph[a].push_back(b);
		graph[b].push_back(a);
	}

	for (int i = 1;i <=N;i++) {
		sort(graph[i].begin(), graph[i].end());
	}

	dfs(V);
	cout << endl;
	bfs(V);
}

File: BaekjoonOnlineJudge/DFS/DFS와BFS (1260)/Solution.py
from collections import deque


def bfs(graph, start, visited):
  visited.append(start)
  queue = deque(graph[start])

  while queue:
    cur = queue.popleft()
    visited.append(cur)
    for node in graph[cur]:
      if node not in visited and node not in queue:
        queue.append(node)
  
  return visited


def dfs(graph, start, visited):
  if start not in visited:
    visited.append(start)
  else:
    return
  for node in graph[start]:
    dfs(graph, node, visited)
  return visited


n, m, v = map(int, input().split())
graph = [[] for i in range(n+1)]

for i in range(1, m+1):
  a, b = map(int, input().split())
  graph[a].extend([b])
  graph[b].extend([a])

graph = list(map(sorted, graph))

print(' '.join(map(str,dfs(graph, v, []))))
print(' '.join(map(str, bfs(graph, v, []))))
